PXC-3567: Merge codership/galera 25.3.32 -> percona/galera 3.x-5.7

venkatesh-prasad-v opened the pull request:
https://jira.percona.com/browse/PXC-3567

Merge tag 'release_25.3.32' into codership_merge_25.3.32_to_3.x-5.7

***
Merge with conflicts resolved.

Conflicts:
        SConstruct

***
Bumped the Galera version to 3.49.

***
Updated GALERA_VERSION_MINOR in the GALERA_VERSION file.

***
Updated GALERA-REVISION to reflect the upstream refresh revision
commit id: 0ef0d79

***
Fix test failures

1. This reverts the code changes in the `gcache/src/GCache_seqno.cpp`
   from "Do not attempt to release above seqno_locked in
   GCache::seqno_release()"
   commit id: c80c268 which was
   introduced in the 25.3.32 release.

2. This backports commit id: e18fc73 (PXC-3024: Node is unable to come
   up if it crashes) and refactored the fix in GCache_seqno to work
   without the now removed condition variable

**Testing Done (with Codership 5.7.32 refresh branch)**
1. Galera and sysvar suites:
Jenkins: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/103/testReport/
Parameters: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/103/parameters/
```
Failing Tests: 

1. galera.galera_sst_xtrabackup-v2_encrypt_with_key - Fails due to PXC-3508
2. galera.galera_sst_xtrabackup-v2_keyring - Fails due to PXC-3508
```
2. FULL MTR:
Jenkins:https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/104/testReport/
Parameters: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/104/parameters/
```
Failing Tests: 
1. binlog.binlog_query_filter_rules - Fixed now
2. keyring_vault tests: https://jira.percona.com/browse/PXC-3297
3. main.merge_mmap & main.merge: https://jira.percona.com/browse/PXC-3296
4. main.commit_1innodb: Sporadic failure due to improper cleanup by its previous test main.bug93770. https://jira.percona.com/browse/PS-7479
5. galera.galera_sst_xtrabackup-v2_encrypt_with_key - Fails due to PXC-3508 
6. galera.galera_sst_xtrabackup-v2_keyring - Fails due to PXC-3508
```

temeo committed:
codership/galera#579 Define aarc64 HWCAP_CRC32 if not defined by libc.

The value for HWCAP_CRC32 on aarch64 may not be defined on some older
platforms.

As the value is part of kernel interface, it is supposed to be stable.

Fixed assertion in crc32c_arm64_tail7().

temeo committed:
CMake files for Galera 3.

Changes in compiled binaries:
- Renamed gcache test program to gcache_test to avoid
  CMake CMP0037 warning.

ayurchen committed:
Not all arm64 devices have hardware CRC32 support. Don't attmept to run
the hardware unit test when not supported.

Refs: codership/galera#582

temeo committed:
Galera RPM spec file for CMake based builds.

temeo committed:
Modified Debian build scripts to use CMake.

temeo committed:
Misc fixes to CMake scripts and packaging:

- Introduced GALERA_VERSION file containinig Galera version information.
- Read Galera version numbers from GALERA_VERSION.
- Create source package using CPack.
- Probe OpenSSL ECDH features.
- Don't run tests in parallel, full output on failure.

temeo committed:
Packaging fixes for Deb/RPM, .gitignore updates.

- Changed from scons to cmake in debian/control.
- No parallel unit tests for RPM build.
- Fixed gcache/src/test -> gcache/src/gcache_test in .gitignore
- Added GALERA_REVISION into .gitignore (generated by cmake/version.cmake).

temeo committed:
Generate galera.spec file with CMake.

- Changed GALERA_REVISION logic to respect input from user.
- Append GALERA_REVISION into package name.

temeo committed:
Improved shared_ptr and unordered_map detection.

temeo committed:
Review fixes:

- Add SSL libraries into required libraries before checking ECDH.
- Set CMP0075 to OLD.
- Enable maintainer mode after feature checks to avoid failures due
  to warnings.

temeo committed:
Renamed generated spec file to codership-galera.spec.

temeo committed:
Review fixes:

- Print CMake version.
- Don't check tmp_ecdh if ecdh_auto check succeeds.
- Fixed CMake version check for OPENSSL_USE_STATIC_LIBS.
- Changed struct _gcache forward declaration to avoid leading
  underscores in type names.
- Reserve vector capacity as the max size is known.

temeo committed:
Option to run unit tests with valgrind.

ayurchen committed:
Initialize SST request buffer in GCS unit tests.
Check SST request length in GCS SST request handler.

Refs codership/galera-bugs#569

temeo committed:
Ensure correct order of linked libraries.

temeo committed:
Added coverage_report target.

temeo committed:
Build improvements: Compiler warnings and CRC32C HW support.

Tuned compiler warnings:

- Removed GALERA_STRICT_BUILD_FLAGS/-Weffc++. This has proven to
  be hard to maintain, better to add warning flags individually.
- Added more warning options when GALERA_MAINTAINER_MODE is enabled.

Fixed some emitted warnings when -Wsign-conversion is enabled:

- Changed gu::ptr_offset to get offset type as template parameter
  to deal with both unsigned and signed offset types.
- Changed flags F_CRC32, F_CRC32C from enum to uin32_t to avoid
  signedness/type width warnings.
- Introduced gu::Buffer::difference_type, use in iterator arithmetics.

Isolated CRC32C HW accelerated code into own library:

- HW acceleration compiler options used only for CRC32C code, avoiding
  accidental use where CPU feature checks have not been implemented.
- Use __get_cpuid() to get CPU feature flags on x86. Function __get_cpuid()
  seems to work at least with GCC and Clang. If __get_cpuid() is not
  supported by compiler, CRC32C HW acceleration will be disabled.

ottok committed:
Clean up and document after CMake introduction

- Update README to state that CMake has now obsoleted Scons
- Update build.sh to automatically use the new CMake style version
  number and thus match the developer experience outlined in the README
- Remove scripts/source as it not used anywhere anymore
- Fix spelling
  - upto -> up to
  - implmented -> implemented

temeo committed:
Minimalistic test program to load/unload wsrep provider.

Test program is independent of unit test framework on purpose
in order to verify that the produced shared library loads
cleanly in minimalistic program.

temeo committed:
Make scripts/build.sh use CMake by default.

- CMake is used by default when scripts/build.sh is run.
- Option --scons can be passed to use old SCons build.

Other cleanups:

- Debian package build using scripts/build.sh --package was removed,
  error is returned if attempted.
- Obsolete scripts/packaging/deb.sh was removed.

ottok committed:
Stop using Debian revision as this is a native upstream package

The Debian revision is intended to show the releases done inside
Debian of the same upstream package. Upstream should not use it.

ottok committed:
Misc fixes to Debian packaging

These fixes have no effect on the resulting package, they are just for
quality assurance and maintainability purposes.

- Run 'wrap-and-sort -av' so all file contents are sorted alphabetically
  and that debian/control file dependencies are listed on separate lines
  to improve diff readability.
- Always use the secure https protocol in all URLs
- Update Debian policy standard
- Change deprecated "Priority: extra" to "optional"
- Fix incomplete license text in debian/copyright
- Ensure build and run-time dependencies are correct
- Sync Debian build hardening flags from downstream

These changes fixes among others the following Lintian warnings:
- vcs-field-uses-insecure-uri
- incomplete-creative-commons-license
- copyright-refers-to-symlink-license
- init.d-script-needs-depends-on-lsb-base
- priority-extra-is-replaced-by-priority-optional

ottok committed:
Update Travis-CI config for CMake

Also ensure build dependencies match the current ones.

ottok committed:
Update URL of wsrep-API sources

temeo committed:
Prefer libcheck_pic if found.

On some older platforms Check library is shipped with two variants,
one build with PIC and one without. If found, prefer check_pic since
all of the code is compiled with PIC option and linking with non-PIC
library may fail.

temeo committed:
RPM and source package build fixes

- Allow overriding Galera version in RPM build.
- Renamed galera-cmake.spec to codership-galera.spec.in
- Bypass full build configuration with GALERA_SOURCE:BOOL=ON to avoid
  requiring all build dependencies in environment where source
  package is generated.

ayurchen committed:
small fixup to build script - don't popd out of build directory until
the build is finished.

ayurchen committed:
Do not attempt to release above seqno_locked in GCache::seqno_release()
Refs codership/galera-bugs#592

temeo committed:
CMake fixes on FreeBSD

- Fixed check library detection.
- Fixed CRC32C detection to check also amd64 system processor.
- Removed FreeBSD specific hacks from os.cmake.

temeo committed:
Bumped version numbers for 25.3.32

temeo committed:
Ensure that GALERA_GIT_REVISION is set

- Read git revision from GALERA_GIT_REVISION.
- If not found, run git rev-parse in CMAKE_SOURCE_DIR to generate
  git revision and write into GALERA_GIT_REVISION.
- Sanity checks for provider_version in wsrep loader test.

Commit:
PXC-3567: Merge codership/galera 25.3.32 -> percona/galera 3.x-5.7

https://jira.percona.com/browse/PXC-3567

Merge tag 'release_25.3.32' into codership_merge_25.3.32_to_3.x-5.7

***
Merge with conflicts resolved.

Conflicts:
        SConstruct

***
Bumped the Galera version to 3.49.

***
Updated GALERA_VERSION_MINOR in the GALERA_VERSION file.

***
Updated GALERA-REVISION to reflect the upstream refresh revision
commit id: 0ef0d79

kamil-holubicki commented on gcache/src/GCache_seqno.cpp:
```diff
@@ -114,8 +114,15 @@ namespace gcache
 
         bool   loop(false);
 
+        bool sleep_a_bit = false;
```
Let's keep the declaration/initialization style consistent in the file.

venkatesh-prasad-v replied:
Fixed.

Commit:
PXC-3567: Merge codership/galera 25.3.32 -> percona/galera 3.x-5.7

https://jira.percona.com/browse/PXC-3567

***
Fix test failures

1. This reverts the code changes in the `gcache/src/GCache_seqno.cpp`
   from "Do not attempt to release above seqno_locked in
   GCache::seqno_release()"
   commit id: c80c268ce871c0b1e5d8c35ba87217c1c5b16d0d which was
   introduced in the 25.3.32 release.

2. This backports commit id: e18fc73c (PXC-3024: Node is unable to come
   up if it crashes) and refactored the fix in GCache_seqno to work
   without the now removed condition variable

kamil-holubicki reviewed (approved):
lgtm
